File: 155.cpp
#include <vector>
#include <stack>
#include <utility>
#include <iostream>

using namespace std;

class MinStack {

private:
    stack<int> min;
    stack<pair<int, int> > s;
    // int min;

public:
    /** initialize your data structure here. */
    MinStack() {
        // min = 0x80000000;
    }
    
    void push(int x) {
        if (s.empty() || getMin() > x) {
            s.push(make_pair(x, x));
        }else {
            s.push(make_pair(x, getMin()));
        }
    }
    
    void pop() {
        if (!s.empty()) s.pop();
    }
    
    int top() {
        return s.top().first;
    }
    
    int getMin() {
        return s.top().second;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

int main(int argc, char const *argv[])
{
    MinStack a;
    a.push(1);
    a.push(2);
    cout << a.getMin();
    return 0;
}

File: 155.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []        

    def push(self, x: int) -> None:
        if not self.stack:
            self.stack.append((x, x))
        else:
            if x < self.stack[-1][1]:
                self.stack.append((x, x))
            else:
                self.stack.append((x, self.stack[-1][1]))

    def pop(self) -> None:
        if self.stack:
            self.stack.pop()

    def top(self) -> int:
        if self.stack:
            return self.stack[-1][0]

    def getMin(self) -> int:
        return self.stack[-1][1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

if __name__ == "__main__":
    obj = MinStack()
    obj.push(1)
    obj.push(2)
    param_3 = obj.top()
    param_4 = obj.getMin()
    print(param_3, param_4)
